### ISSUE TITLE
Message attachments can't be opened on iOS, macOS (and maybe other platforms)

### ISSUE THREAD

**Issue by jbksr** [miDeb/digitales_register#18]
Flutter run log on macOS:
`[log] failed to download file https://osz-schlanders.digitalesregister.it/v2/api/message/messageSubmissionDownloadEntry: PathAccessException: Cannot open file, path = '/Users/user/Library/Containers/com.package/Data/Downloads/msg_22_9_07.09.2023 Mitteilung Eltern.pdf' (OS Error: Operation not permitted, errno = 1)`

Flutter run log on iOS:
`[log] failed to download file https://osz-schlanders.digitalesregister.it/v2/api/message/messageSubmissionDownloadEntry: PathNotFoundException: Cannot open file, path = '/var/mobile/Containers/Data/Application/F0C2B614-0C82-4C09-BBEA-52669AEB7E52/Downloads/msg_22_9_07.09.2023 Mitteilung Eltern.pdf' (OS Error: No such file or directory, errno = 2)`

**Comment by jbksr** [comment]
Seems to fail on these lines in `middleware.dart`. Permission issue?

```
final sink = saveFile.openWrite();
await sink.addStream((result.data as dio.ResponseBody).stream);
await sink.flush();
await sink.close();
```

**Comment by jbksr** [comment]
```
Future<bool> downloadFile(
  String url,
  String fileName,
  Map<String, dynamic> parameters,
) async {
  await wrapper.ensureLoggedIn();

  File saveFile = File(
    "${await _getAttachmentDownloadDirectory()}/$fileName",
  );

  // TODO: Temporary fix for iOS
  if (Platform.isIOS) {
    saveFile = File(
      "${(await getApplicationDocumentsDirectory()).path}/$fileName",
    );
  }

  ...
}
```

This could work?

**Comment by miDeb** [comment]
I think the problem is that the downloads directory does not exist yet, we have to create it before being able to use it. I will have a look later today.

**Comment by jbksr** [comment]
```
Future<String> _getAttachmentDownloadDirectory() async {
  try {
    Directory? directory;

    if (Platform.isIOS || Platform.isMacOS) {
      directory = await getApplicationDocumentsDirectory();
    } else {
      directory = await getDownloadsDirectory();
    }

    if (directory != null) {
      return directory.path;
    }
  } catch (_) {}
  return (await getApplicationDocumentsDirectory()).path;
}
```

This seems to work both for iOS and macOS. But I'm worried this might break something for devices running older iOS and macOS versions.

**Comment by jbksr** [comment]
Also worth mentioning is that the Play Store Android version has a similar behaviour (does not open the file). But for some weird reason it works just fine when debugging.

**Comment by jbksr** [comment]
![Screenshot_2023-09-25-07-08-46-543_com android chrome-edit](https://github.com/miDeb/digitales_register/assets/73539478/af79cf15-5948-44b2-83bb-bb27ea8d6172)
[docs.flutter.dev](https://docs.flutter.dev/cookbook/persistence/reading-writing-files) 

This Flutter Docs states

**Comment by jbksr** [comment]
Sorry for that above, accidentally pressed the wrong button.

![Screenshot_2023-09-25-07-08-46-543_com android chrome-edit](https://github.com/miDeb/digitales_register/assets/73539478/07240801-5c4b-45d5-b425-ae3b0a176068)
[docs.flutter.dev](https://docs.flutter.dev/cookbook/persistence/reading-writing-files) 

This Flutter docs article states that we should use the documents directory for persisting user data.
```
Future<String> get _localPath async {
  final directory = await getApplicationDocumentsDirectory();

  return directory.path;
}
```

I'll try to migrate the `_getAttachmentDownloadDirectory` function to use this instead of the `getDownloadsDirectory` and then test it on different devices.

**Comment by miDeb** [comment]
No worries :) Let me know when you're stuck on something, I should also have more time in the holidays to respond much quicker

**Comment by miDeb** [comment]
Did you test the flatpak version for linux? I don't know if it behaves differently because it is supposed to be a sandbox and not allow programs to access user files. But it shouldn't be a problem because if I remember correctly the file will just be downloaded again if it no longer exists. Anyway, most users are probably using Android or iOS, so it's important those platforms work, and only a few are on macOS, Windows and even fewer on Linux. But I appreciate your testing... I have to admit I usually only tested one or two platforms when making changes like this.

**Comment by miDeb** [comment]
>  Anyways, I've compiled it for Windows, it launches but I can't log in. It says there's no internet connection. I've also tried the existing version directly from the Microsoft Store and still the same thing. So unfortunately I wasn't able to test that.
"Keine Verbindung mit "..." möglich. Bitte überprüfe deine Internetverbindung. Wenn du ..."

I couldn't reproduce this on my machine on a fresh install (I get the correct user_not_found error from the server), so it's hopefully not a bug with the app.

**Comment by miDeb** [comment]
Closing this, as hopefully this has now been fixed by #20.
